File: bench/suite/batch/python/batch.py
"""heavy/batch in Python: multiprocessing over newline-aligned byte ranges
of an mmap, dicts per process merged at the end. See bench/SPEC.md."""
import mmap
import multiprocessing as mp
import os
import sys

BLOCK = 16 * 1024 * 1024


def work(args):
    path, start, end = args
    regions = {}
    users = {}
    skus = {}
    rows = 0
    ug = users.get
    sg = skus.get
    with open(path, "rb") as f:
        mm = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ)
        pos = start
        while pos < end:
            stop = min(pos + BLOCK, end)
            if stop < end:
                stop = mm.find(b"\n", stop - 1) + 1 or end
            for line in mm[pos:stop].splitlines():
                _ts, user, sku, qty, price, region = line.split(b",")
                qty = int(qty)
                rev = qty * int(price)
                r = regions.get(region)
                if r is None:
                    r = regions[region] = [0, 0, 0]
                r[0] += 1
                r[1] += qty
                r[2] += rev
                user = int(user)
                users[user] = ug(user, 0) + rev
                skus[sku] = sg(sku, 0) + rev
                rows += 1
            pos = stop
        mm.close()
    return rows, regions, users, skus


def main(path):
    size = os.path.getsize(path)
    workers = int(os.environ.get("WORKERS") or len(os.sched_getaffinity(0)) if hasattr(os, "sched_getaffinity") else os.cpu_count())
    workers = min(workers, size // 65536 + 1)
    bounds = [0]
    with open(path, "rb") as f:
        mm = mmap.mmap(f.fileno(), 0, access=mmap.ACCESS_READ) if size else None
        for w in range(1, workers):
            at = size // workers * w
            nl = mm.find(b"\n", at - 1)
            bounds.append(max(bounds[-1], nl + 1 if nl >= 0 else size))
        if mm:
            mm.close()
    bounds.append(size)
    ranges = [(path, bounds[i], bounds[i + 1]) for i in range(workers)]
    with mp.get_context("fork").Pool(workers) as pool:
        parts = pool.map(work, ranges)

    rows = 0
    regions = {}
    users, skus = {}, {}
    for r, reg, u, s in parts:
        rows += r
        for code, (c, q, v) in reg.items():
            t = regions.setdefault(code, [0, 0, 0])
            t[0] += c
            t[1] += q
            t[2] += v
        if not users:
            users, skus = u, s
            continue
        ug, sg = users.get, skus.get
        for k, v in u.items():
            users[k] = ug(k, 0) + v
        for k, v in s.items():
            skus[k] = sg(k, 0) + v

    import heapq
    top_users = heapq.nsmallest(100, users.items(), key=lambda kv: (-kv[1], kv[0]))
    top_skus = heapq.nsmallest(10, skus.items(), key=lambda kv: (-kv[1], kv[0]))
    out = [f"rows={rows}"]
    for code in sorted(regions):
        c, q, v = regions[code]
        out.append(f"region={code.decode()} count={c} qty={q} revenue={v}")
    for i, (u, v) in enumerate(top_users, 1):
        out.append(f"top_user rank={i} user_id={u} revenue={v}")
    for i, (s, v) in enumerate(top_skus, 1):
        out.append(f"top_sku rank={i} sku={s.decode()} revenue={v}")
    sys.stdout.write("\n".join(out) + "\n")


if __name__ == "__main__":
    main(sys.argv[1])

File: bench/suite/batch/js/batch.mjs
// heavy/batch for Node and Bun: worker_threads over newline-aligned byte
// ranges read with positional fs.readSync, Maps per worker merged at the
// end. The same file runs on both runtimes. See bench/SPEC.md.
import { Worker, isMainThread, parentPort, workerData } from "node:worker_threads";
import fs from "node:fs";
import os from "node:os";

const BLOCK = 16 * 1024 * 1024;

function work({ path, start, end }) {
  const fd = fs.openSync(path, "r");
  const regions = new Float64Array(676 * 3); // count, qty, revenue per code
  const users = new Map();
  const skus = new Map();
  let rows = 0;
  let buf = Buffer.allocUnsafe(BLOCK + 256);
  let carry = 0;
  let pos = start;
  while (pos < end) {
    const want = Math.min(BLOCK, end - pos);
    if (carry + want > buf.length) {
      const nb = Buffer.allocUnsafe(carry + want);
      buf.copy(nb, 0, 0, carry);
      buf = nb;
    }
    const got = fs.readSync(fd, buf, carry, want, pos);
    pos += got;
    const len = carry + got;
    let last = len - 1;
    while (last >= 0 && buf[last] !== 10) last--;
    let i = 0;
    while (i <= last) {
      while (buf[i] !== 44) i++;
      i++;
      let user = 0;
      while (buf[i] !== 44) user = user * 10 + (buf[i++] - 48);
      i++;
      const skuStart = i;
      while (buf[i] !== 44) i++;
      const sku = buf.latin1Slice(skuStart, i);
      i++;
      let qty = 0;
      while (buf[i] !== 44) qty = qty * 10 + (buf[i++] - 48);
      i++;
      let price = 0;
      while (buf[i] !== 44) price = price * 10 + (buf[i++] - 48);
      i++;
      const r = ((buf[i] - 65) * 26 + (buf[i + 1] - 65)) * 3;
      i += 3;
      const rev = qty * price;
      regions[r] += 1;
      regions[r + 1] += qty;
      regions[r + 2] += rev;
      rows++;
      users.set(user, (users.get(user) || 0) + rev);
      skus.set(sku, (skus.get(sku) || 0) + rev);
    }
    carry = len - (last + 1);
    buf.copy(buf, 0, last + 1, len);
  }
  fs.closeSync(fd);
  return { rows, regions, users: [...users], skus: [...skus] };
}

if (!isMainThread) {
  parentPort.postMessage(work(workerData));
} else {
  const path = process.argv[2];
  const size = fs.statSync(path).size;
  let workers = Number(process.env.WORKERS) || os.availableParallelism();
  workers = Math.min(workers, Math.floor(size / 65536) + 1);

  const fd = fs.openSync(path, "r");
  const probe = Buffer.allocUnsafe(4096);
  const bounds = [0];
  for (let w = 1; w < workers; w++) {
    let at = Math.floor(size / workers) * w;
    let found = size;
    for (let p = at - 1; p < size;) {
      const got = fs.readSync(fd, probe, 0, 4096, p);
      if (got === 0) break;
      const nl = probe.subarray(0, got).indexOf(10);
      if (nl >= 0) { found = p + nl + 1; break; }
      p += got;
    }
    bounds.push(Math.max(found, bounds[bounds.length - 1]));
  }
  bounds.push(size);
  fs.closeSync(fd);

  const parts = await Promise.all(bounds.slice(0, -1).map((start, w) => new Promise((resolve, reject) => {
    const worker = new Worker(new URL(import.meta.url), { workerData: { path, start, end: bounds[w + 1] } });
    worker.once("message", resolve);
    worker.once("error", reject);
  })));

  let rows = 0;
  const regions = new Float64Array(676 * 3);
  const users = new Map();
  const skus = new Map();
  for (const p of parts) {
    rows += p.rows;
    for (let k = 0; k < regions.length; k++) regions[k] += p.regions[k];
    for (const [u, v] of p.users) users.set(u, (users.get(u) || 0) + v);
    for (const [s, v] of p.skus) skus.set(s, (skus.get(s) || 0) + v);
  }

  const topRev = [], topUser = [];
  for (const [u, v] of users) {
    let n = topRev.length;
    if (n === 100 && !(v > topRev[99] || (v === topRev[99] && u < topUser[99]))) continue;
    if (n === 100) { topRev.pop(); topUser.pop(); n--; }
    let at = n;
    while (at > 0 && (v > topRev[at - 1] || (v === topRev[at - 1] && u < topUser[at - 1]))) at--;
    topRev.splice(at, 0, v);
    topUser.splice(at, 0, u);
  }
  const topSkus = [...skus].sort((a, b) => b[1] - a[1] || (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0)).slice(0, 10);

  const lines = [`rows=${rows}`];
  for (let c = 0; c < 676; c++) {
    if (regions[c * 3] > 0) {
      const code = String.fromCharCode(65 + Math.floor(c / 26), 65 + (c % 26));
      lines.push(`region=${code} count=${regions[c * 3]} qty=${regions[c * 3 + 1]} revenue=${regions[c * 3 + 2]}`);
    }
  }
  topRev.forEach((v, i) => lines.push(`top_user rank=${i + 1} user_id=${topUser[i]} revenue=${v}`));
  topSkus.forEach(([s, v], i) => lines.push(`top_sku rank=${i + 1} sku=${s} revenue=${v}`));
  process.stdout.write(lines.join("\n") + "\n");
}
